sdks/python/apache_beam/transforms/enrichment_handlers/bigquery.py

```diff
@@ -0,0 +1,265 @@
+from google.cloud import bigquery
+
+import apache_beam as beam
+from apache_beam.pvalue import Row
+from apache_beam.transforms.enrichment import EnrichmentSourceHandler
+
+QueryFn = Callable[[beam.Row], str]
+ConditionValueFn = Callable[[beam.Row], List[Any]]
+
+
+def _validate_batch_query_fn(query_fn, min_batch_size, max_batch_size):
+  if query_fn and min_batch_size and max_batch_size:
+    raise ValueError(
+        "Please provide exactly one of `query_fn` or "
+        "(`min_batch_size` and `max_batch_size`)")
+
+
+def _validate_bigquery_metadata(
+    table_name, row_restriction_template, fields, condition_value_fn, query_fn):
+  if query_fn and bool(table_name or row_restriction_template or fields or
+                       condition_value_fn):
+    raise ValueError(
+        "Please provide either `query_fn` or the parameters "
+        "`table_name`, `row_restriction_template`, and `fields` "
+        "together.")
+  elif not query_fn and not (table_name and row_restriction_template and
+                             (fields or condition_value_fn)):
+    raise ValueError(
+        "Please provide either `query_fn` or the parameters "
+        "`table_name`, `row_restriction_template`, and"
+        "`fields/condition_value_fn` together.")
+  if not query_fn and ((fields and condition_value_fn) or
+                       (not fields and not condition_value_fn)):
+    raise ValueError(
+        "Please provide exactly one of `fields` or "
+        "`condition_value_fn`")
+
+
+class BigQueryEnrichmentHandler(EnrichmentSourceHandler[Union[Row, List[Row]],
+                                                        Union[Row, List[Row]]]):
+  """Enrichment handler for Google Cloud BigQuery.
+
+  Use this handler with :class:`apache_beam.transforms.enrichment.Enrichment`
+  transform.
+
+  To use this handler you need either of the following combinations:
+    * `table_name`, `row_restriction_template`, `fields`
+    * `table_name`, `row_restriction_template`, `condition_value_fn`
+    * `query_fn`
+
+  By default, the handler pulls all columns from the BigQuery table.
+  To override this, use the `column_name` parameter to specify a list of column
+  names to fetch.
+
+  This handler pulls data from BigQuery per element by default. To change this
+  behavior, set the `min_batch_size` and `max_batch_size` parameters.
+  These min and max values for batch size are sent to the
+  :class:`apache_beam.transforms.utils.BatchElements` transform.
+
+  NOTE: Elements cannot be batched when using the `query_fn` parameter.
+  """
+  def __init__(
+      self,
+      project: str,
+      *,
+      table_name: str = "",
+      row_restriction_template: str = "",
+      fields: Optional[List[str]] = None,
+      column_names: Optional[List[str]] = None,
+      condition_value_fn: Optional[ConditionValueFn] = None,
+      query_fn: Optional[QueryFn] = None,
+      min_batch_size: Optional[int] = None,
+      max_batch_size: Optional[int] = None,
```

Comment:
Can we default this to `10,000` and remove it if `query_fn` is specified? I think we want batching to happen by default in the `table_name` case

Reply:
done